### PR TITLE
Add `ephys_precluster` module

### PR DESCRIPTION
**Pull request by kabilar**
- Add tables for running multiple pre-clustering analysis steps on the acquired data.
- First I created a copy of `ephys_acute` within `ephys_precluster` and then added the additional tables.
- The `PreClusterParamStep` table allows for keeping track of multiple pre-clustering steps and their order of operations.
- Load the pre-clustered data.
- Data populates without error.
- Update changelog and version.
- Add documentation on ephys modules.
- Add images for `ephys_chronic` and `ephys_precluster` schemas.
- Future work
    - https://github.com/datajoint/workflow-array-ephys/issues/67
    - Trigger the pre-clustering steps

### CODE REVIEW

**Comment by kabilar** [comment]
@Alvalunasan

**Comment by kabilar** [comment]
Hi @Alvalunasan, I couldn't add you as a reviewer but please feel free to add any feedback.  Thanks.

**Comment by kabilar** [comment]
Noticed an issue with the implementation.  Will address and then mark this pull request as `Ready for review`.

**Comment by kabilar** [comment]
Hi @ttngu207, I have fixed the issue that I noted above in this [commit](https://github.com/datajoint/element-array-ephys/pull/65/commits/3d8ec16a85507d4708ceedfdcffd4609546dc0f7).

**Comment by kabilar** [comment]
Hi @sidhulyalkar, as you are learning more about `element-array-ephys`.  Feel free to review this pull request.  Thanks.

File: element_array_ephys/ephys_precluster.py [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+    # Ordered list of paramset_idx that are to be run
+    precluster_param_list_id: smallint
+    """
+
+    class ParamOrder(dj.Part):
+        definition = """
+        -> master
+        order_id: smallint                  # Order of operations
+        ---
+        -> [nullable] PreClusterParamSet    # Nullable for when pre-clustering is not performed.
```
Comment (kabilar) [inline element_array_ephys/ephys_precluster.py:305]:
```suggestion
        -> PreClusterParamSet
```

File: element_array_ephys/ephys_precluster.py [hunk 2]
```diff
@@ -0,0 +1,892 @@
...
+                    'The specified param-set'
+                    ' already exists - paramset_idx: {}'.format(existing_paramset_idx))
+        else:
+            cls.insert1(param_dict)
+
+
+@schema
+class PreClusterParamList(dj.Manual):
+    definition = """
+    # Ordered list of paramset_idx that are to be run
```
Comment (kabilar) [inline element_array_ephys/ephys_precluster.py:296]:
```suggestion
    # Ordered list of paramset_idx that are to be run
    #  When pre-clustering is not performed, do not create an entry in `ParamOrder`
```

File: element_array_ephys/ephys_precluster.py [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+    precluster_param_list_id: smallint
+    """
+
+    class ParamOrder(dj.Part):
```
Comment (kabilar) [inline element_array_ephys/ephys_precluster.py:300]:
```suggestion
    class Step(dj.Part):
```

File: element_array_ephys/ephys_precluster.py [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+        definition = """
+        -> master
+        order_id: smallint                  # Order of operations
```
Comment (kabilar) [inline element_array_ephys/ephys_precluster.py:303]:
```suggestion
        order: smallint                  # Order of operations
```

File: element_array_ephys/ephys_precluster.py [hunk 5]
```diff
@@ -0,0 +1,892 @@
...
+    package_version='': varchar(16)
+    """
+
+    def make(self, key):
+        task_mode, output_dir = (PreClusterTask & key).fetch1('task_mode', 
+                                                              'precluster_output_dir')
+        precluster_output_dir = find_full_path(get_ephys_root_data_dir(), output_dir)
+
+        if task_mode == 'none':
+            creation_time = (EphysRecording & key).fetch1('recording_datetime')
```
Comment (kabilar) [inline element_array_ephys/ephys_precluster.py:345]: Throw an error if `task_mode=none` and there are entries in `PreClusterParamStep.Step`.

Comment (kabilar) [reply element_array_ephys/ephys_precluster.py:349]: Added `ValueError`.

**Comment by kabilar** [comment]
Hi @ttngu207, I have implemented the changes we discussed and updated the first comment to reflect all changes.  This pull request is ready for review.  Thanks!

File: element_array_ephys/ephys_precluster.py [hunk 6]
```diff
@@ -0,0 +1,896 @@
...
+        if param_query:  # If the specified param-set already exists
+            existing_paramset_idx = param_query.fetch1('paramset_idx')
+            if existing_paramset_idx == paramset_idx:  # If the existing set has the same paramset_idx: job done
+                return
+            else:  # If not same name: human error, trying to add the same paramset with different name
+                raise dj.DataJointError(
+                    'The specified param-set'
+                    ' already exists - paramset_idx: {}'.format(existing_paramset_idx))
+        else:
+            cls.insert1(param_dict)
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:290]:
```suggestion
        cls.insert1(param_dict)
```

You do not need to explicitly handle the duplication errors. DataJoint will already throw a reasonable exception.

File: element_array_ephys/ephys_precluster.py [hunk 7]
```diff
@@ -0,0 +1,896 @@
...
+    def make(self, key):
+
+        session_dir = find_full_path(get_ephys_root_data_dir(), 
+                                     get_session_directory(key))
+
+        inserted_probe_serial_number = (ProbeInsertion * probe.Probe & key).fetch1('probe')
+
+        # search session dir and determine acquisition software
+        for ephys_pattern, ephys_acq_type in zip(['*.ap.meta', '*.oebin'],
+                                                 ['SpikeGLX', 'Open Ephys']):
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:154]:
```suggestion
        for ephys_pattern, ephys_acq_type in (('*.ap.meta', 'SpikeGLX'), ('*.oebin',  'Open Ephys')):
```

File: element_array_ephys/ephys_precluster.py (hunk 6 above)
Comment (dimitri-yatsenko) [reply element_array_ephys/ephys_precluster.py:290]: If the datajoint exception is not clear enough, we can work to improve it.

File: element_array_ephys/ephys_precluster.py [hunk 8]
```diff
@@ -0,0 +1,896 @@
...
+        if param_query:  # If the specified param-set already exists
+            existing_paramset_idx = param_query.fetch1('paramset_idx')
+            if existing_paramset_idx == paramset_idx:  # If the existing set has the same paramset_idx: job done
+                return
+            else:  # If not same name: human error, trying to add the same paramset with different name
+                raise dj.DataJointError(
+                    'The specified param-set'
+                    ' already exists - paramset_idx: {}'.format(existing_paramset_idx))
+        else:
+            cls.insert1(param_dict)
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:517]:
```suggestion
        cls.insert1(param_dict)
```

DataJoint already handles these duplication errors better than this code.

File: element_array_ephys/ephys_precluster.py [hunk 9]
```diff
@@ -0,0 +1,896 @@
...
+    """
+
+    def create1_from_clustering_task(self, key, curation_note=''):
+        """
+        A function to create a new corresponding "Curation" for a particular 
+        "ClusteringTask"
+        """
+        if key not in Clustering():
+            raise ValueError(f'No corresponding entry in Clustering available'
+                             f' for: {key}; do `Clustering.populate(key)`')
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:602]: This is a foreign key check.  Database already does this better than this code.
```suggestion
```

File: element_array_ephys/ephys_precluster.py [hunk 10]
```diff
@@ -0,0 +1,896 @@
...
+                spike_count = len(unit_spike_times)
+
+                units.append({
+                    'unit': unit,
+                    'cluster_quality_label': unit_lbl,
+                    **channel2electrodes[unit_channel],
+                    'spike_times': unit_spike_times,
+                    'spike_count': spike_count,
+                    'spike_sites': spike_sites[kilosort_dataset.data['spike_clusters'] == unit],
+                    'spike_depths': spike_depths[kilosort_dataset.data['spike_clusters'] == unit]})
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:685]:
```suggestion
        for unit, unit_label in zip(valid_units, valid_unit_labels):
            if (kilosort_dataset.data['spike_clusters'] == unit).any():
                unit_channel, _ = kilosort_dataset.get_best_channel(unit)
                unit_spike_times = (spike_times[kilosort_dataset.data['spike_clusters'] == unit]
                                    / kilosort_dataset.data['params']['sample_rate'])
                spike_count = len(unit_spike_times)

                units.append({
                    'unit': unit,
                    'cluster_quality_label': unit_label,
                    **channel2electrodes[unit_channel],
                    'spike_times': unit_spike_times,
                    'spike_count': spike_count,
                    'spike_sites': spike_sites[kilosort_dataset.data['spike_clusters'] == unit],
                    'spike_depths': spike_depths[kilosort_dataset.data['spike_clusters'] == unit]})
```

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

File: element_array_ephys/ephys_precluster.py [hunk 11]
```diff
@@ -0,0 +1,896 @@
...
+
+    @classmethod
+    def insert_new_params(cls, precluster_method: str, paramset_idx: int,
+                          paramset_desc: str, params: dict):
+        param_dict = {'precluster_method': precluster_method,
+                      'paramset_idx': paramset_idx,
+                      'paramset_desc': paramset_desc,
+                      'params': params,
+                      'param_set_hash':  dict_to_uuid(params)}
+        param_query = cls & {'param_set_hash': param_dict['param_set_hash']}
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:279]: This `param_query` is unnecessary

File: element_array_ephys/ephys_precluster.py [hunk 12]
```diff
@@ -0,0 +1,896 @@
...
+        elif acq_software == 'Open Ephys':
+            
+            session_dir = find_full_path(get_ephys_root_data_dir(), 
+                                         get_session_directory(key))
+
+            loaded_oe = openephys.OpenEphys(session_dir)
+            oe_probe = loaded_oe.probes[probe_sn]
+
+            lfp_channel_ind = np.arange(
+                len(oe_probe.lfp_meta['channels_ids']))[-1::-self._skip_channel_counts]
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:442]:
```suggestion
            lfp_channel_ind = slice(len(oe_probe.lfp_meta['channels_ids']))-1, 0, -self._skip_channel_counts)
```

File: element_array_ephys/ephys_precluster.py [hunk 13]
```diff
@@ -0,0 +1,896 @@
...
+                              lfp_time_stamps=lfp_timestamps,
+                              lfp_mean=lfp.mean(axis=0)))
+
+            electrode_query = (probe.ProbeType.Electrode
+                               * probe.ElectrodeConfig.Electrode
+                               * EphysRecording & key)
+            probe_electrodes = {key['electrode']: key
+                                for key in electrode_query.fetch('KEY')}
+
+            for channel_idx in np.array(oe_probe.lfp_meta['channels_ids'])[lfp_channel_ind]:
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:459]:
```suggestion
            for channel_idx in oe_probe.lfp_meta['channels_ids'][lfp_channel_ind]:
```

File: element_array_ephys/ephys_precluster.py [hunk 14]
```diff
@@ -0,0 +1,896 @@
...
+                electrode_keys.append(probe_electrodes[channel_idx])
+        else:
+            raise NotImplementedError(f'LFP extraction from acquisition software'
+                                      f' of type {acq_software} is not yet implemented')
+
+        # single insert in loop to mitigate potential memory issue
+        for electrode_key, lfp_trace in zip(electrode_keys, lfp):
+            self.Electrode.insert1({**key, **electrode_key, 'lfp': lfp_trace})
```
Comment (dimitri-yatsenko) [inline element_array_ephys/ephys_precluster.py:467]:
```suggestion
        self.Electrode.insert(dict(key, lfp=lfp_trace, **electrode_key) 
                    for electrode_key, lfp_trace in zip(electrode_keys, lfp))
```

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

**Comment by kabilar** [comment]
Thanks @dimitri-yatsenko!  To expedite this merge and since this pull request is focused on adding the `PreCluster` tables, I will create GitHub issues to address your comments in a future pull request.  Your suggestions will have to be replicated across `ephys_acute`, `ephys_chronic`, `ephys_precluster`, and `ephys_no_curation`.

**Review by ttngu207 — APPROVED** [review]
Perhaps add a bit of context/story - e.g. this version of the element-array-ephys is being deployed and used by the Princeton U19 program

**Review by dimitri-yatsenko — APPROVED** [review]

**Comment by dimitri-yatsenko** [comment]
@kabilar No need for issues. I will issue a PR with these changes.

**Pull request merged**